File: OJ/leet342/leet342.cpp
/*
leet342


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    bool isPowerOfFour(int n) {
        cout << (n & (-n)) << endl;
        if (not (n >= 1 and (n & (-n)) == n)) return false;
        auto cnt = 0;
        while (n) {
            cnt++; n >>= 1;
        }
        return cnt % 2 == 1;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << boolalpha;
    cout << sol.isPowerOfFour(1) << endl;
    cout << sol.isPowerOfFour(4) << endl;
    cout << sol.isPowerOfFour(16) << endl;
    return 0;
}

File: OJ/leet342/leet342.py
# leet342

from typing import List


class Solution:
    # 要点 x & (x - 1) 判断是否是 2 的幂次, 4^k % 3 == 1, 4^k * 2 % 3 == 2
    def isPowerOfFour1(self, x: int) -> bool:
        return  any((x == (1<<i) for i in range(0, 32, 2)) )
    def isPowerOfFour(self, x: int) -> bool:
        return  (x & (x-1)) == 0 and (x % 3 == 1)


print(Solution().isPowerOfFour(1))
print(Solution().isPowerOfFour(1<<2))
print(Solution().isPowerOfFour(1<<16))
print(Solution().isPowerOfFour(1<<30))
print(Solution().isPowerOfFour(1<<32))


